src/poetry/puzzle/solver.py

```diff
@@ -164,12 +164,13 @@ def _solve(
             if package.features:
                 for _package in packages:
                     if (
-                        _package.name == package.name
-                        and not _package.is_same_package_as(package)
+                        not _package.features
+                        and package.provides(_package)
                         and _package.version == package.version
                     ):
                         for dep in package.requires:
-                            if dep.is_same_package_as(_package):
+                            # Prevent adding base package as a dependency to itself
+                            if _package.provides(dep):
```

Comment:
@radoering @dimbleby PTAL -- I can see from reading the code that this is indeed correct given the current implementation. However, I'm wondering if this is another issue with `is_same_package_as` being brittle, or just a misuse. If I understand the recent discussion and refactoring correctly, this is just a misuse and broke as a result of fixing the overly-aggressive merging of packages.

Reply:
LGTM. This is a place where we don't care for `source_type`, so clearly a misuse of `is_same_package_as`.

Maybe, we can change L179-L182 to use `else` instead of `continue`. Imho, that improves readability.

Reply:
> Maybe, we can change L179-L182 to use `else` instead of `continue`. Imho, that improves readability.

Done